File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:

        ans = 0
        current_sum =0 
        begin = 0
        end = 0
        nums_set = {}

        while end < len(nums):
            curr_num = nums[end]
            last_occurence = nums_set.get(curr_num, -1)

            while begin <= last_occurence or end - begin + 1 > k:
                current_sum -= nums[begin]
                begin += 1

            nums_set[curr_num] = end
            current_sum += nums[end]
            if end - begin + 1 == k:
                ans = max(ans, current_sum)
            end += 1
        return ans

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int  begin = 0, end =0;
        long ans = 0, currentSum = 0;
        
        HashMap<Integer, Integer> numSet = new HashMap<Integer, Integer>();

        while(end< nums.length){
            int currentNum = nums[end];
            int lastOccurence = numSet.getOrDefault(currentNum, -1);

            while(begin <= lastOccurence || end - begin + 1> k){
                currentSum -= nums[begin];
                begin += 1;  
            }

            currentSum += nums[end];
            numSet.put(currentNum, end);
            if(end - begin + 1 == k){
                ans = Math.max(ans, currentSum);
            }
            end += 1;
        }

        return ans;
    }
}